Arch arm mpu assert additions and fixes

ioannisg opened the pull request:
Addresses #7384.
Addresses an additional bug related to region out-of-bounds configuration (see relevant comment).
Adds some minor fixes to the code and comment enhancements.

ioannisg committed:
arch: arm: coding and comment style fixes

This commit fixes some minor issues with coding style
and comment syntax in arm_mpu.c

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: enhance internal function documentation

This commit enhaces the documentation of internal functions
in arm_mpu.c by explicitly stating that the caller needs to
ensure the validity of the supplied MPU region index. The
warning is required as these functions modify the ARM MPU_RNR
register, without checking themselves the validity of the
provided region number.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: implement internal function to disable MPU region

This commit adds an internal function to disable an MPU region.
The function includes an assert that the requested MPU region
number is a valid one. arm_mpu.c is refactor to use this
function in all cases where an MPU region needs to be disabled.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: add additional sanity checks before MPU config change

This commit inserts sanity checks every time we are performing
a (re)-configuration of one or multiple MPU regions, ensuring
that we do not attempt to configure an invalid region number.

Particulary for arm_mpu_config(), called during pre-kernel
initialization phase, we add a system ASSERT if we attempt
to initialize more regions that what is supported by hardware.
We do this to ensure the misconfiguration is detected early and
the system boot is aborted.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -273,8 +273,7 @@ void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
 	u32_t region_attr = _get_region_attr_by_type(type, size);
 
-	/* ARM MPU supports up to 16 Regions */
-	if (region_index > _get_num_regions()) {
```
This looks like an out-of-bounds sanity-check bug, @agross-linaro, am I correct?

tyrantlizard77 reviewed (changes requested):
I think this code does too much.

It's sufficient to just make assertion checks, not try to prevent the system from crashing if assertions are off and people feed garbage data to the MPU driver. We only do this kind of rigorous runtime checking for drivers that we expose to user mode.

tyrantlizard77 reviewed (approved):
I think this is fine for 1.12

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -267,11 +302,14 @@ void arm_core_mpu_configure_user_context(struct k_thread *thread)
-	if (size > 0)
-		_region_init(index, (u32_t)&__app_ram_start, region_attr);
-#endif
+	if (index < _get_num_regions())
```
Thanks!

MaureenHelm commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -143,7 +143,7 @@ static inline u32_t _get_region_index_by_type(u32_t type)
 
 /**
- * This internal function check if region is enabled or not
+ * This internal function checks if region is enabled or not.
```
Can you please make the same comment fixes in the nxp mpu?

ioannisg replied:
Will have a look now.

MaureenHelm reviewed (commented):
LGTM. Some of the comments you're fixing are also in the nxp mpu, could you please fix those too?

ioannisg committed:
arch: arm: mpu: minor comment style fixes

This commit fixes some minor function documentation issues
and comments' style in the NXP_MPU driver.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@MaureenHelm  I pushed an additional commit that implements the comment&style fixes in NXP_MPU.
